### ISSUE TITLE
how to software encoding

### ISSUE THREAD

**Issue by FrankLove** [pedroSG94/RootEncoder#1258]
hi, i find my devices use this library to push  stream when first open the camera preview will stuck about 2s, did the library already realize software encoding,how can i use it?can you help me?
 Force H264 and AAC Codec hardware/software encoding (Not recommended).

**Comment by pedroSG94** [comment]
You can use setForce method. Like this:
```java
rtmpCamera1.setForce(CodecUtil.Force.SOFTWARE, CodecUtil.Force.SOFTWARE);
```
The first paramater is for video encoding and the second for audio.

**Comment by FrankLove** [comment]
Thanks for your reply,But there's one thing I don't quite understand,It takes approximately 3 seconds  while opening the camera preview on low-performance devices，but when i using the system's default camera app or writing my own camera preview code, it is likely that are better and It takes approximately 1.5 seconds ，why is this?Can I optimize the camera preview opening time on low-performance devices？

**Comment by pedroSG94** [comment]
The time about open camera is because you are not only opening the camera:
You are also creating microphone and both encoders. You can create a thread to load all in background to avoid screen freeze but make sure that you call methods in a sync way.

About resolution:
You are right. The higher resolution supported is the preview resolution using camera1 (using camera2 you can use others resolutions) because you haven't other way to extract raw frames to encode with camera1 in other way.

**Comment by FrankLove** [comment]
thanks for your replay.
About the first question According  on my test and understand ,on the on high-performance devices we can't find the difference  when open the camera preview ,but on low-performance devices ,will see it ,the TextureView mode to open preview is take more  time than surfaceveiw mode  although　i use a thread for textrueview,because my device can't support camera2,so I can only use cammera1.
on my devices when i change your source code this method（getMaxEncoderSizeSupported） to this，now it can push the stream of 1920 * 1080,but i don't is right to do this?but it reached the effect I wanted
```java 
private Camera.Size getMaxEncoderSizeSupported(Camera.Parameters parameters, Camera.Size previewSize) {
    float previewRatio = (float) previewSize.width / previewSize.height;
    List<Camera.Size> supportedPictureSizes = parameters.getSupportedPictureSizes();
    Camera.Size maxSupportedSize = null;
    for (Camera.Size size : supportedPictureSizes) {
        float pictureRatio = (float) size.width / size.height;

        if (Math.abs(pictureRatio - previewRatio) < 0.1 && size.width >= previewSize.width) {
            if (maxSupportedSize == null || (size.width * size.height) > (maxSupportedSize.width * maxSupportedSize.height)) {
                maxSupportedSize = size;
            }
        }
    }
    return maxSupportedSize;
}`

**Comment by pedroSG94** [comment]
Normaly getSupportedPictureSizes return a list of sizes for image capture.
I think that with that method you are basically skipping the method to filter resolutions supported by the encoder.
Can you try this?:

Change the method to:
```java
private Camera.Size getMaxEncoderSizeSupported(int cameraId) {
    if (CamcorderProfile.hasProfile(cameraId, CamcorderProfile.QUALITY_2160P)) {
      return camera.new Size(3840, 2160);
    } else if (CamcorderProfile.hasProfile(cameraId, CamcorderProfile.QUALITY_1080P)) {
      return camera.new Size(1920, 1080);
    } else if (CamcorderProfile.hasProfile(cameraId, CamcorderProfile.QUALITY_720P)) {
      return camera.new Size(1280, 720);
    } else {
      return camera.new Size(640, 480);
    }
  }
```
Use cameraSelected as parameter:
<img width="459" alt="image" src="https://github.com/pedroSG94/rtmp-rtsp-stream-client-java/assets/17556876/39fdc97d-8e10-454b-a130-1f9e30d24893">

Let me know the result please.

**Comment by FrankLove** [comment]
thanks for your reply ,when i use your new method on my devices to test,it returns the result 
return camera.new Size(1920, 1080);
and the old method is return 
return camera.new Size(640, 480);
so it seems your new method it more rigorous ,can you will change your source code to that?

**Comment by pedroSG94** [comment]
Ok, perfect
I will change it today

**Comment by pedroSG94** [comment]
You can use this gradle:
```gradle
implementation "com.github.pedroSG94.rtmp-rtsp-stream-client-java:rtplibrary:cd8a6c5d6b"
```
This is a compilation of the commit that contain the change. That commit will be included in 2.2.7 version but i will update it later

**Comment by FrankLove** [comment]
thanks,wonderful ！
